Update to world events

Fusezion opened the pull request:
### Description
<!--- Describe your changes here. --->
This PR adds `of %worlds%` into the `WorldSaveEvent`, `WorldLoadEvent`, `WorldUnloadEvent`, and `WorldInitEvent`
I've also added a new effect for manually saving worlds
```applescript
save all worlds
save [the] %worlds%
```

---
**Target Minecraft Versions:** <!-- 'any' means all supported versions --> any
**Requirements:** <!-- Required plugins, Minecraft versions, server software... --> N/A
**Related Issues:** <!-- Links to related issues --> #5113 #3656

TheLimeGlass commented:
I don't think you need `world` before %worlds% as the ClassInfo parse will remove the word world

Fusezion commented:
Alright I'll remove it in a bit once I finish the breedable stuff

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldSave.java:
```diff
@@ -0,0 +1,63 @@
+public class EvtWorldSave extends SkriptEvent {
+
+	static {
+		Skript.registerEvent("World Save", EvtWorldSave.class, WorldSaveEvent.class, "world sav(e|ing) [of %worlds%]")
```
```suggestion
		Skript.registerEvent("World Save", EvtWorldSave.class, WorldSaveEvent.class, "world sav(e|ing) [of %-worlds%]")
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldSave.java:
```diff
@@ -0,0 +1,63 @@
+
+	@Override
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		if (worlds != null) worlds = ((Literal<World>) args[0]).getArray();
```
```suggestion
		if (worlds != null)
			worlds = ((Literal<World>) args[0]).getArray();
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldSave.java:
```diff
@@ -0,0 +1,63 @@
+
+	@Override
+	public boolean check(Event event) {
+		if (worlds == null) return true;
```
```suggestion
		if (worlds == null)
			return true;
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldUnload.java:
```diff
@@ -0,0 +1,62 @@
+public class EvtWorldUnload extends SkriptEvent {
+
+	static {
+		Skript.registerEvent("World Unload", EvtWorldUnload.class, WorldUnloadEvent.class, "world unload[ing] [of %worlds%]")
```
```suggestion
		Skript.registerEvent("World Unload", EvtWorldUnload.class, WorldUnloadEvent.class, "world unload[ing] [of %-worlds%]")
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldUnload.java:
```diff
@@ -0,0 +1,62 @@
+
+	static {
+		Skript.registerEvent("World Unload", EvtWorldUnload.class, WorldUnloadEvent.class, "world unload[ing] [of %worlds%]")
+			.description("Called when a world is unloaded. This event might never be called if you don't have a world management plugin.")
```
```suggestion
			.description("Called when a world is unloaded. This event will never be called if you don't have a world management plugin.")
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldUnload.java:
```diff
@@ -0,0 +1,62 @@
+
+	@Override
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		if (worlds != null) worlds = ((Literal<World>) args[0]).getArray();
```
```suggestion
		if (worlds != null) worlds =
			((Literal<World>) args[0]).getArray();
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldUnload.java:
```diff
@@ -0,0 +1,62 @@
+
+	@Override
+	public boolean check(Event event) {
+		if (worlds == null) return true;
```
```suggestion
		if (worlds == null)
			return true;
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldUnload.java:
```diff
@@ -0,0 +1,62 @@
+
+
+	@Override
+	public String toString(final @Nullable Event event, final boolean debug) {
```
```suggestion
	public String toString(@Nullable Event event, boolean debug) {
```

Fusezion commented on src/main/java/ch/njol/skript/effects/EffWorldUnload.java:
```diff
@@ -0,0 +1,85 @@
+		for (World world : this.world.getArray(event)) {
+			if (mainWorld != null) {
+				World finalMainWorld = mainWorld;
+				world.getPlayers().forEach(player -> player.teleport(finalMainWorld.getSpawnLocation()));
```
After a quick test of this removed I've discerned that this is not how it works in fact if the player is currently within said world they aren't disconnected nor teleported to another world, instead the world will stay loaded and never unload nor send an unload event

Fusezion commented on src/main/java/ch/njol/skript/effects/EffWorldUnload.java:
```diff
@@ -0,0 +1,85 @@
+	@Override
+	protected void execute(Event event) {
+		World mainWorld = Bukkit.getWorlds().get(0);
+		if (world == mainWorld) {
```
I've decided to in my next push to remove the `.get(1)` stuff and only default to the first world in `.get(0)` I believe this is a better solution as the first world should be the world you've labeled main anyways and should never be unloaded

TheLimeGlass commented on src/main/java/ch/njol/skript/effects/EffWorldUnload.java:
```diff
@@ -0,0 +1,78 @@
+	@Override
+	protected void execute(Event event) {
+		World mainWorld = Bukkit.getWorlds().get(0);
+		Bukkit.broadcastMessage(mainWorld.toString());
```
Debug message

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldInit.java:
```diff
@@ -0,0 +1,66 @@
+
+	@Override
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		if (worlds != null)
+			worlds = ((Literal<World>) args[0]).getArray();
```
the worlds variable is always going to be null. You need to set it to the argument, same with others
```suggestion
		worlds = ((Literal<World>) args[0]).getArray();
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldLoad.java:
```diff
@@ -0,0 +1,66 @@
+	@Override
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		if (worlds != null)
+			worlds = ((Literal<World>) args[0]).getArray();
```
I stated this for the other event, but worlds is always going to be null so there is no point of the `if (worlds != null)` statement.

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldSave.java:
```diff
@@ -0,0 +1,66 @@
+	@Override
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		if (worlds != null)
+			worlds = ((Literal<World>) args[0]).getArray();
```
I stated this for the other event, but worlds is always going to be null so there is no point of the if (worlds != null) statement.

Fusezion replied:
Since removing this in events I've received this error from it
``java.lang.NullPointerException: Cannot invoke "ch.njol.skript.lang.Literal.getArray()" because "args[0]" is null``

this occurs when it's `on world save:`

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldUnload.java:
```diff
@@ -0,0 +1,66 @@
+	@Override
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		if (worlds != null)
+			worlds = ((Literal<World>) args[0]).getArray();
```
I stated this for the other event, but worlds is always going to be null so there is no point of the if (worlds != null) statement.

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorldInit.java:
```diff
@@ -0,0 +1,65 @@
+			.since("1.0, INSERT VERSION (defining worlds)");
+	}
+
+	@Nullable
```
Actually I think this doesn't need to be nullable. If you set the `[of %-worlds%]` to `[of %worlds%]` it should grab the event-value world of the event, and then it'll just be checking the event world with the event world, meaning a gurenteed pass, similar to return true. so the null check in the check(event) methods can be removed and the nullable annotations on these events.

Fusezion replied:
From the short little test I've done it doesn't seem do want to do that, instead when the check is called `worlds` is just null and causes a runtime error, probably is best to be left alone, or I'm doing something wrong

Fusezion replied:
From the testing I've been able to do with this I was unable to get it to work, so it's possible I've done something wrong knowing me but for now I think it's fine I could always revisit it in the future.

TheLimeGlass commented on src/main/java/ch/njol/skript/effects/EffWorldUnload.java:
```diff
@@ -0,0 +1,75 @@
+	protected void execute(Event event) {
+		World mainWorld = Bukkit.getWorlds().get(0);
+		for (World world : this.worlds.getArray(event)) {
+			if (mainWorld != null && world != mainWorld) {
+				world.getPlayers().forEach(player -> player.teleport(mainWorld.getSpawnLocation()));
+			}
```
```suggestion
			if (mainWorld != null && world != mainWorld)
				world.getPlayers().forEach(player -> player.teleport(mainWorld.getSpawnLocation()));
```

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffWorldSave.java:
```diff
@@ -0,0 +1,67 @@
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
```
Should go after `execute`

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffWorldSave.java:
```diff
@@ -0,0 +1,67 @@
+
+	@Override
+	protected void execute(Event event) {
+		for (World world : worlds.getArray(event)) {
+			world.save();
+		}
```
```suggestion
		for (World world : worlds.getArray(event))
			world.save();
```

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffWorldSave.java:
```diff
@@ -0,0 +1,67 @@
+	"save all worlds"
+})
+@Since("INSERT VERSION")
+public class EffWorldSave extends Effect {
```
Can saving many worlds at once freeze the server? Might be worth mentioning if so

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffWorldUnload.java:
```diff
@@ -0,0 +1,74 @@
+public class EffWorldUnload extends Effect {
+
+	static {
+		Skript.registerEffect(EffWorldUnload.class, "unload %worlds% [and (:save)]");
```
```suggestion
		Skript.registerEffect(EffWorldUnload.class, "unload %worlds% [:without saving]");
```
I think this syntax is better

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffWorldUnload.java:
```diff
@@ -0,0 +1,74 @@
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "unload world(s) " + worlds.toString(event, debug) + " " + (save ? "with saving" : "without saving");
+	}
```
```suggestion
	}

```

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffWorldUnload.java:
```diff
@@ -0,0 +1,74 @@
+		for (World world : this.worlds.getArray(event)) {
+			if (mainWorld != null && world != mainWorld)
+				world.getPlayers().forEach(player -> player.teleport(mainWorld.getSpawnLocation()));
+			Bukkit.unloadWorld(world, save);
```
What happens if you unload a world containing players?

Fusezion replied:
I remember originally having a talk somewhere about this, if players are in the world 1 of 2 things happen. they are teleported to the main world or it just fails to unload

since I remember testing this in the `mainWorld` and it just didn't do anything didn't teleport me to another world or unload the world, so I think I did this as a backup safe check

Fusezion replied:
For above, I've revisited this and it seems to be working fine without the teleport portion, unloads `mainWorld` teleports me to one of the others, unload that world resends me to the `mainWorld`. Seems like `mainWorld` still can't be unloaded but will atleast teleport players outside of it

APickledWalrus commented on src/main/java/ch/njol/skript/events/EvtWorldInit.java:
```diff
@@ -0,0 +1,64 @@
+	@Override
+	@Nullable
+	public String toString(Event event, boolean debug) {
+		return "initialization of world" + (worlds == null ? "" : " of " + worlds.toString(event,debug));
```
```suggestion
		return "world initialization" + (worlds == null ? "" : " of "_ + worlds.toString(event,debug));
```

APickledWalrus commented on src/main/java/ch/njol/skript/events/EvtWorldLoad.java:
```diff
@@ -0,0 +1,65 @@
+	@Override
+	@Nullable
+	public String toString( Event event, boolean debug) {
+		return "loading of world" + (worlds == null ? "" : " of " + worlds.toString(event,debug));
```
```suggestion
		return "world loading" + (worlds == null ? "" : " of " + worlds.toString(event,debug));
```

APickledWalrus commented on src/main/java/ch/njol/skript/events/EvtWorldLoad.java:
```diff
@@ -0,0 +1,65 @@
+	public boolean check(Event event) {
+		if (worlds == null)
+			return true;
+
```
```suggestion
```

APickledWalrus commented on src/main/java/ch/njol/skript/events/EvtWorldUnload.java:
```diff
@@ -0,0 +1,65 @@
+	public boolean check(Event event) {
+		if (worlds == null)
+			return true;
+
```
```suggestion
```

APickledWalrus commented on src/main/java/ch/njol/skript/events/EvtWorldUnload.java:
```diff
@@ -0,0 +1,65 @@
+	@Override
+	@Nullable
+	public String toString(Event event, boolean debug) {
+		return "unloading of world" + (worlds == null ? "" : " of " + worlds.toString(event,debug));
```
```suggestion
		return "world unloading" + (worlds == null ? "" : " of " + worlds.toString(event,debug));
```

APickledWalrus commented on src/main/java/ch/njol/skript/events/EvtWorldSave.java:
```diff
@@ -0,0 +1,65 @@
+	public boolean check(Event event) {
+		if (worlds == null)
+			return true;
+
```
```suggestion
```

APickledWalrus commented on src/main/java/ch/njol/skript/events/EvtWorldSave.java:
```diff
@@ -0,0 +1,65 @@
+	@Override
+	@Nullable
+	public String toString(Event event, boolean debug) {
+		return "saving of world" + (worlds == null ? "" : " of " + worlds.toString(event,debug));
```
```suggestion
		return "world saving" + (worlds == null ? "" : " of " + worlds.toString(event,debug));
```

APickledWalrus reviewed (changes requested):
Nice additions :)

Instead of having four event classes though, I think you should just do something like https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/events/EvtItem.java where multiple events are registered in the same file (so that they use the same check method) since these check methods are very similar

AyhamAl-Ali commented on src/main/java/ch/njol/skript/effects/EffWorldUnload.java:
```diff
@@ -0,0 +1,72 @@
+import org.eclipse.jdt.annotation.Nullable;
+
+@Name("Unload World")
+@Description({"Unload a given world with optional saving"})
```
```suggestion
@Description("Unload a given world with optional saving")
```

AyhamAl-Ali commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,100 @@
+		if (worlds != null) {
+			boolean isAnd = worlds.getAnd();
+			if (isAnd) {
+				Skript.error("An \"and\" list cannot be used in this event as multiple worlds cannot load at the same time", ErrorQuality.SEMANTIC_ERROR);
```
```suggestion
				Skript.error("An \"and\" list cannot be used in this event as multiple worlds cannot load at the same time");
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,99 @@
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		worlds = (Literal<World>) args[0];
+		if (worlds != null) {
+			boolean isAnd = worlds.getAnd();
```
Doesn't need to be placed in a boolean variable, you can use it in the condition directly

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,99 @@
+		if (worlds != null) {
+			boolean isAnd = worlds.getAnd();
+			if (isAnd) {
+				Skript.error("An \"and\" list cannot be used in this event as multiple worlds cannot load at the same time");
```
The event is not always a load world, so the error message is incorrect when it comes to another event that's not a load world.

Fusezion replied:
this seem like a better error?
`"An \"and\" list cannot be used in this event as multiple worlds cannot be called at the same time"`

Fusezion commented:
Was asked to add a method to load the world, will be working on it and hopefully should be committed later today or early tomorrow

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,98 @@
+	@Override
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		worlds = (Literal<World>) args[0];
+		if (worlds != null) {
+			if (worlds.getAnd()) {
```
```diff
+		if (worlds != null && worlds.getAnd()) {
```

Fusezion commented:
> Was asked to add a method to load the world, will be working on it and hopefully should be committed later today or early tomorrow
> 
> **EDIT:** this will actually have to wait


After mentioning the issue exists within SkBee to Shane he kindly went and looked into the issues
in order for us a safely add a `load world %string%` effect we'd have to add a portion to explicitly state the world environment
meaning it would have to look something like `load [world] %string% with [environment] %environment%` to load existing `non-normal` worlds correctly. I'll work on adding this into the PR once I'm home from work today

TheLimeGlass commented:
`with [environment] %environment%` should be optional

TheLimeGlass commented on src/main/java/ch/njol/skript/effects/EffWorldLoad.java:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	private boolean save, load;
+	private Expression<Object> worlds;
```
```suggestion
	private Expression<?> worlds;
```

TheLimeGlass commented on src/main/java/ch/njol/skript/effects/EffWorldLoad.java:
```diff
@@ -0,0 +1,97 @@
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		worlds = (Expression<Object>) exprs[0];
```
```suggestion
		worlds = exprs[0];
```

TheLimeGlass commented on src/main/java/ch/njol/skript/effects/EffWorldLoad.java:
```diff
@@ -0,0 +1,97 @@
+				if (environment != null)
+					worldCreator.environment(environment);
+				worldCreator.createWorld();
+			}
+			else if (!load && world instanceof World) {
```
```suggestion
			} else if (!load && world instanceof World) {
```

TheLimeGlass commented on src/main/java/ch/njol/skript/effects/EffWorldLoad.java:
```diff
@@ -0,0 +1,97 @@
+	@Nullable
+	private Expression<Environment> environment;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
```
```suggestion
	@Override
	@SuppressWarnings("unchecked")
	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
```

TheLimeGlass commented on src/main/java/ch/njol/skript/effects/EffWorldSave.java:
```diff
@@ -0,0 +1,69 @@
+
+	private Expression<World> worlds;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
```
```suggestion
	@Override
	@SuppressWarnings("unchecked")
	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,96 @@
+	static {
+		// World Save Event
+		Skript.registerEvent("World Save", EvtWorld.class, WorldSaveEvent.class, "world sav(e|ing) [of %-worlds%]")
+			.description("Called when a world is saved to disk. Usually all worlds are saved simultaneously, but world management plugins could change this.")
+			.examples(
+				"on world save of \"world\":",
+				"\tbroadcast \"The world %event-world% has been saved\"")
+			.since("1.0, INSERT VERSION (defining worlds)");
```
```suggestion
				.description("Called when a world is saved to disk. Usually all worlds are saved simultaneously, but world management plugins could change this.")
				.examples(
					"on world save of \"world\":",
					"\tbroadcast \"The world %event-world% has been saved\"")
				.since("1.0, INSERT VERSION (defining worlds)");
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,96 @@
+
+		// World Init Event
+		Skript.registerEvent("World Init", EvtWorld.class, WorldInitEvent.class, "world init[ialization] [of %-worlds%]")
+			.description("Called when a world is initialized. As all default worlds are initialized before",
+				"any scripts are loaded, this event is only called for newly created worlds.",
+				"World management plugins might change the behaviour of this event though.")
+			.examples("on world init of \"world_the_end\":")
+			.since("1.0, INSERT VERSION (defining worlds)");
```
```suggestion
				.description("Called when a world is initialized. As all default worlds are initialized before",
					"any scripts are loaded, this event is only called for newly created worlds.",
					"World management plugins might change the behaviour of this event though.")
				.examples("on world init of \"world_the_end\":")
				.since("1.0, INSERT VERSION (defining worlds)");
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,96 @@
+
+		// World Unload Event
+		Skript.registerEvent("World Unload", EvtWorld.class, WorldUnloadEvent.class, "world unload[ing] [of %-worlds%]")
+			.description("Called when a world is unloaded. This event will never be called if you don't have a world management plugin.")
+			.examples(
+				"on world unload:",
+				"\tbroadcast \"the %event-world% has been unloaded!\"")
+			.since("1.0, INSERT VERSION (defining worlds)");
```
```suggestion
				.description("Called when a world is unloaded. This event will never be called if you don't have a world management plugin.")
				.examples(
					"on world unload:",
					"\tbroadcast \"the %event-world% has been unloaded!\"")
				.since("1.0, INSERT VERSION (defining worlds)");
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,96 @@
+
+		// World Load Event
+		Skript.registerEvent("World Load", EvtWorld.class, WorldLoadEvent.class, "world load[ing] [of %-worlds%]")
+			.description("Called when a world is loaded. As with the world init event, this event will not be called for the server's default world(s).")
+			.examples(
+				"on world load of \"world_nether\":",
+				"\tbroadcast \"The world %event-world% has been loaded!\"")
+			.since("1.0, INSERT VERSION (defining worlds)");
```
```suggestion
				.description("Called when a world is loaded. As with the world init event, this event will not be called for the server's default world(s).")
				.examples(
					"on world load of \"world_nether\":",
					"\tbroadcast \"The world %event-world% has been loaded!\"")
				.since("1.0, INSERT VERSION (defining worlds)");
```

TheLimeGlass commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,96 @@
+	@Nullable
+	private Literal<World> worlds;
+
+	@Override
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
```
```suggestion
	@Override
	@SuppressWarnings("unchecked")
	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
```

sovdeeth commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,97 @@
+	@SuppressWarnings("unchecked")
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		worlds = (Literal<World>) args[0];
+		if (worlds != null && worlds.getAnd()) {
+			Skript.error("An \"and\" list cannot be used in this event as multiple worlds cannot be called at the same time");
+			return false;
+		}
```
This prevents you from doing `on world load of "world_nether"` because it's an and list of 1, I think.
![image](https://github.com/SkriptLang/Skript/assets/10354869/78b2859b-9d03-4963-9deb-60b49a397dd6)
![image](https://github.com/SkriptLang/Skript/assets/10354869/a2f3cfc3-2891-4db2-97f8-ef072ade4606)

sovdeeth replied:
![image](https://github.com/SkriptLang/Skript/assets/10354869/2644a931-8c5c-4d5a-9948-795e677707a9)
A solution could be what I did for evtgrow, where you force And lists to be Or.

Fusezion replied:
> ![image](https://github.com/SkriptLang/Skript/assets/10354869/2644a931-8c5c-4d5a-9948-795e677707a9)
> A solution could be what I did for evtgrow, where you force And lists to be Or.

Wonder if something like that is better for skript to eventually force. Iirc my set and wasn't needed as internally it's an or

Fusezion replied:
So am back at this for first thing of the new year and I'm honestly unsure if I should bother keeping this error. For one `and/or` both work wording wise fine to me. The whole reason it's even done is `on save of "World" and "World_Nether"` being read as called when both worlds are saved rather than one or the other but skript doesn't ever take this into account.

![image](https://github.com/SkriptLang/Skript/assets/81952476/bda48074-7731-46e9-bcd1-a47d3e1830b9)

Fusezion replied:
Alright so after a bit of looking, I decided to do what EvtClick does which should of been the suggested path tbh. When the defined areas are single i.e. `%world%` instead of `%worlds%` you're still able to do patterns like `world save [of %world%]` which allows `world` and `world or world_nether` will error with `world and world_the_end` without me handling it.

New commits will be done shortly

sovdeeth replied:
> For one `and/or` both work wording wise fine to me. 

Yes that was my suggestion, which I did for evtgrow. You allow the user to write either, but you act as if it is OR either way.

Fusezion commented:
Latest commit remove the `worlds` plurality and the suggested `getAnd` error message. this event still works with plurals however only as `on world save of "world" or "world_the_end` if you attempt to use `and` skript will treat it as a list and error with the base one. 

If a change is wanted to add it back I have zero intention to 'getAnd' calls `on world save of "world"` an and list when it should be false. Skript will need to change that behavior before this is attempted again.

sovdeeth commented:
> Latest commit remove the `worlds` plurality and the suggested `getAnd` error message. this event still works with plurals however only as `on world save of "world" or "world_the_end` if you attempt to use `and` skript will treat it as a list and error with the base one. 
> 
> If a change is wanted to add it back I have zero intention to 'getAnd' calls `on world save of "world"` an and list when it should be false. Skript will need to change that behavior before this is attempted again.

Sorry I'm really confused about what you're trying to say here
All I understood is that you fixed the issue with a single world failing?

Fusezion commented:
> Sorry I'm really confused about what you're trying to say here All I understood is that you fixed the issue with a single world failing?

Fixed? yes, lazily to be fair. I basically just removed it, Skript treats `1 or 2` the same as `1` so `%world%` supports both chpoces without `1 and 2` support. But I believe I might have miss understood your review, was I thought you were against `and` being a supported option

Fusezion commented:
Alright now that I understood you weren't against `and` being supported, I've reverted previous commit and added your suggested of `setAnd(false)` let me know if I did it wrong

Fusezion commented:
Alright so updated my branch and it broke it as `setAnd` was renamed in #6000 to `invertAnd`

sovdeeth commented on src/main/java/ch/njol/skript/events/EvtWorld.java:
```diff
@@ -0,0 +1,96 @@
+	@SuppressWarnings("unchecked")
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		worlds = (Literal<World>) args[0];
+		if (worlds instanceof LiteralList<?>) {
```
```suggestion
		if (worlds.getAnd() && worlds instanceof LiteralList<?>) {
```

sovdeeth reviewed (changes requested):
One small change

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffWorldLoad.java:
```diff
@@ -0,0 +1,97 @@
+public class EffWorldLoad extends Effect {
+
+	static {
+		Skript.registerEffect(EffWorldLoad.class, "load [[the] world] %strings% [with environment %-environment%]", "unload [[the] world] %worlds% [:without saving]");
```
```suggestion
		Skript.registerEffect(EffWorldLoad.class,
				"load [[the] world[s]] %strings% [with environment %-environment%]",
				"unload [[the] world[s]] %worlds% [:without saving]"
		);
```

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffWorldLoad.java:
```diff
@@ -0,0 +1,97 @@
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		if (load)
+			return "load world(s) " + worlds.toString(event, debug) + (environment == null ? "" : " with environment " + environment.toString(event, debug));
+		return "unload world(s) " + worlds.toString(event, debug) + " " + (save ? "with saving" : "without saving");
```
```suggestion
			return "load the world(s) " + worlds.toString(event, debug) + (environment == null ? "" : " with environment " + environment.toString(event, debug));
		return "unload the world(s) " + worlds.toString(event, debug) + " " + (save ? "with saving" : "without saving");
```

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffWorldSave.java:
```diff
@@ -0,0 +1,70 @@
+public class EffWorldSave extends Effect {
+
+	static {
+		Skript.registerEffect(EffWorldSave.class, "save %worlds%");
```
```suggestion
		Skript.registerEffect(EffWorldSave.class, "save [[the] world[s]] %worlds%");
```

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffWorldSave.java:
```diff
@@ -0,0 +1,70 @@
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "save world(s) " + worlds.toString(event, debug);
```
```suggestion
		return "save the world(s) " + worlds.toString(event, debug);
```

APickledWalrus reviewed (changes requested):
just some concerns about syntax consistency
